tests/spsc/priority_queue.cpp

```diff
@@ -66,6 +67,103 @@ TEST_CASE("Write multiple with different priority and read back ensuring "
     REQUIRE(read == 1024);
 }
 
+TEST_CASE("Multithreaded read/write", "[pq_multithreaded]") {
+    lockfree::spsc::PriorityQueue<uint64_t, 10, 4> queue;
+    std::vector<std::thread> threads;
+    std::vector<uint64_t> written;
+    std::vector<uint64_t> read;
+
+    /* The following code pushes a large number of values into a priority
+       queue using four different priorities. The priority is also encoded
+       into the value pushed, as its two lower significance bits.
+       
+       Both consumer and producer store in two vectors the numbers written
+       and read.
+
+       After the multi-threaded execution, special code (described in detail
+       below) checks no higher priority value was written to the priority
+       queue at the time a lower priority one was read.
+    */
+
+    // consumer, it just pops values from the queue and stores them in the
+    // main thread vector
+    threads.emplace_back([&]() {
+        uint64_t value = 0;
+        uint64_t cnt = 0;
+        do {
+            bool pop_success = queue.Pop(value);
+            if (pop_success) {
+              read.push_back(value);
+              cnt++;
+            }
+        } while (cnt < TEST_MT_TRANSFER_CNT);
+    });
+
+    // producer, uses alternative priorities and pushes a counter shifted to
+    // accommodate the priority on its lower bits.
+    threads.emplace_back([&]() {
+        uint64_t cnt = 0;
+        uint64_t value = 0;
+        uint8_t prio = 0;
+        do {
+          value = cnt << 2 + prio;
+          bool push_success = queue.Push(value, prio);
+          if (push_success) {
+            written.push_back(value);
+            prio = (prio + 1) % 4; // this could be also randomly generated
+            cnt++;
+          }
+        } while (cnt < TEST_MT_TRANSFER_CNT + 1);
+    });
+    for (auto &t : threads) {
+        t.join();
+    }
+    /* The following code checks that at all times no higher priority value was
+       present in the `written` vector.
+
+       It needs to keep track which values were already read, it does that with
+       the help of the `consumed` Boolean vector.
+
+       As an optimization, the variable `start` contains the longest prefix of
+       `consumed` that is all true.
+    */
+    std::vector<bool> consumed(written.size(), false);    
+    uint64_t value1, value2;
+    uint8_t prio1, prio2;
+    size_t start = 0;
```

Comment:
Can we remove the start/all_consumed optimization entirely? As this is a test we don't really care about performance as much as readability.

Reply:
Sure thing. Just note that running time from the test goes from less than 1 second to about 15 minutes.

Reply:
I could not reproduce this amount of slowdown, for me it ran quite fast locally, and the GitHub workflow finished in under a minute.